File: samples/CameraAccess/CameraAccess/OpenClaw/ToolCallModels.swift
import Foundation

// MARK: - Gemini Tool Call (parsed from server JSON)

struct GeminiFunctionCall {
  let id: String
  let name: String
  let args: [String: Any]
}

struct GeminiToolCall {
  let functionCalls: [GeminiFunctionCall]

  init?(json: [String: Any]) {
    guard let toolCall = json["toolCall"] as? [String: Any],
          let calls = toolCall["functionCalls"] as? [[String: Any]] else {
      return nil
    }
    self.functionCalls = calls.compactMap { call in
      guard let id = call["id"] as? String,
            let name = call["name"] as? String else { return nil }
      let args = call["args"] as? [String: Any] ?? [:]
      return GeminiFunctionCall(id: id, name: name, args: args)
    }
  }
}

// MARK: - Gemini Tool Call Cancellation

struct GeminiToolCallCancellation {
  let ids: [String]

  init?(json: [String: Any]) {
    guard let cancellation = json["toolCallCancellation"] as? [String: Any],
          let ids = cancellation["ids"] as? [String] else {
      return nil
    }
    self.ids = ids
  }
}

// MARK: - Tool Result

enum ToolResult {
  case success(String)
  case failure(String)

  var responseValue: [String: Any] {
    switch self {
    case .success(let result):
      return ["result": result]
    case .failure(let error):
      return ["error": error]
    }
  }
}

// MARK: - Tool Call Status (for UI)

enum ToolCallStatus: Equatable {
  case idle
  case executing(String)
  case completed(String)
  case failed(String, String)
  case cancelled(String)

  var displayText: String {
    switch self {
    case .idle: return ""
    case .executing(let name): return "Running: \(name)..."
    case .completed(let name): return "Done: \(name)"
    case .failed(let name, let err): return "Failed: \(name) - \(err)"
    case .cancelled(let name): return "Cancelled: \(name)"
    }
  }

  var isActive: Bool {
    if case .executing = self { return true }
    return false
  }
}

// MARK: - Tool Declarations (for Gemini setup message)

enum ToolDeclarations {

  static func allDeclarations() -> [[String: Any]] {
    return [execute]
  }

  /// The single tool. The model sets `action` so the app can decide between
  /// delegating a real-world task to OpenClaw and rendering widgets in the UI.
  static let execute: [String: Any] = [
    "name": "execute",
    "description": "Your single tool for taking any action. First DECIDE the kind of action and set \"action\":\n• action=\"render\": display visual widget cards in the user's field of view — provide \"widgets\".\n• action=\"delegate\": perform a real-world action (sending messages, searching, lists, reminders, notes, scheduling, smart-home control, app interactions, research, etc.) — provide a detailed \"task\". This is the only way anything real happens.",
    "parameters": [
      "type": "object",
      "properties": [
        "action": [
          "type": "string",
          "enum": ["delegate", "render"],
          "description": "delegate = perform a real-world action via the assistant (OpenClaw); render = show widget cards in the field of view.",
        ],
        "task": [
          "type": "string",
          "description": "For action=delegate: a clear, detailed description of what to do, with all relevant context (names, content, platforms, quantities, etc.).",
        ],
        "widgets": [
          "type": "array",
          "description": "For action=render: the widget cards to show, top to bottom. Each render replaces whatever is currently shown.",
          "items": [
            "type": "object",
            "properties": [
              "type": [
                "type": "string",
                "enum": ["text", "image", "table"],
                "description": "The widget type.",
              ],
              "title": [
                "type": "string",
                "description": "Optional heading shown at the top of the widget.",
              ],
              "body": [
                "type": "string",
                "description": "For type=text: the paragraph of text to show.",
              ],
              "imageKind": [
                "type": "string",
                "enum": ["map", "gallery", "calendar"],
                "description": "For type=image: which fixed image to show — map (a location/Google map), gallery (a photo gallery), or calendar (a calendar).",
              ],
              "caption": [
                "type": "string",
                "description": "For type=image: a short caption under the image.",
              ],
              "columns": [
                "type": "array",
                "items": ["type": "string"],
                "description": "For type=table: optional column header labels.",
              ],
              "rows": [
                "type": "array",
                "description": "For type=table: the rows; each row is an array of cell strings.",
                "items": [
                  "type": "array",
                  "items": ["type": "string"],
                ],
              ],
            ] as [String: Any],
            "required": ["type"],
          ] as [String: Any],
        ] as [String: Any],
      ] as [String: Any],
      "required": ["action"],
    ] as [String: Any],
    "behavior": "BLOCKING",
  ]
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/openclaw/ToolCallModels.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw

import org.json.JSONArray
import org.json.JSONObject

// Gemini Tool Call (parsed from server JSON)

data class GeminiFunctionCall(
    val id: String,
    val name: String,
    val args: Map<String, Any?>
)

data class GeminiToolCall(
    val functionCalls: List<GeminiFunctionCall>
) {
    companion object {
        fun fromJSON(json: JSONObject): GeminiToolCall? {
            val toolCall = json.optJSONObject("toolCall") ?: return null
            val calls = toolCall.optJSONArray("functionCalls") ?: return null
            val functionCalls = mutableListOf<GeminiFunctionCall>()
            for (i in 0 until calls.length()) {
                val call = calls.getJSONObject(i)
                val id = call.optString("id", "")
                val name = call.optString("name", "")
                if (id.isEmpty() || name.isEmpty()) continue
                val argsObj = call.optJSONObject("args")
                val args = mutableMapOf<String, Any?>()
                if (argsObj != null) {
                    for (key in argsObj.keys()) {
                        args[key] = argsObj.opt(key)
                    }
                }
                functionCalls.add(GeminiFunctionCall(id, name, args))
            }
            return if (functionCalls.isNotEmpty()) GeminiToolCall(functionCalls) else null
        }
    }
}

// Gemini Tool Call Cancellation

data class GeminiToolCallCancellation(
    val ids: List<String>
) {
    companion object {
        fun fromJSON(json: JSONObject): GeminiToolCallCancellation? {
            val cancellation = json.optJSONObject("toolCallCancellation") ?: return null
            val idsArray = cancellation.optJSONArray("ids") ?: return null
            val ids = mutableListOf<String>()
            for (i in 0 until idsArray.length()) {
                ids.add(idsArray.getString(i))
            }
            return if (ids.isNotEmpty()) GeminiToolCallCancellation(ids) else null
        }
    }
}

// Tool Result

sealed class ToolResult {
    data class Success(val result: String) : ToolResult()
    data class Failure(val error: String) : ToolResult()

    fun toJSON(): JSONObject = when (this) {
        is Success -> JSONObject().put("result", result)
        is Failure -> JSONObject().put("error", error)
    }
}

// Tool Call Status (for UI)

sealed class ToolCallStatus {
    data object Idle : ToolCallStatus()
    data class Executing(val name: String) : ToolCallStatus()
    data class Completed(val name: String) : ToolCallStatus()
    data class Failed(val name: String, val error: String) : ToolCallStatus()
    data class Cancelled(val name: String) : ToolCallStatus()

    val displayText: String
        get() = when (this) {
            is Idle -> ""
            is Executing -> "Running: $name..."
            is Completed -> "Done: $name"
            is Failed -> "Failed: $name - $error"
            is Cancelled -> "Cancelled: $name"
        }

    val isActive: Boolean
        get() = this is Executing
}

// OpenClaw Connection State

sealed class OpenClawConnectionState {
    data object NotConfigured : OpenClawConnectionState()
    data object Checking : OpenClawConnectionState()
    data object Connected : OpenClawConnectionState()
    data class Unreachable(val message: String) : OpenClawConnectionState()
}

// Tool Declarations (for Gemini setup message)

object ToolDeclarations {
    fun allDeclarationsJSON(): JSONArray {
        return JSONArray().put(executeJSON())
    }

    private fun executeJSON(): JSONObject {
        return JSONObject().apply {
            put("name", "execute")
            put(
                "description",
                "Your single tool for taking any action. First DECIDE the kind of action and set \"action\":\n" +
                    "• action=\"render\": display visual widget cards in the user's field of view — provide \"widgets\".\n" +
                    "• action=\"delegate\": perform a real-world action (sending messages, searching, lists, reminders, notes, scheduling, smart-home control, app interactions, research, etc.) — provide a detailed \"task\". This is the only way anything real happens.",
            )
            put("parameters", JSONObject().apply {
                put("type", "object")
                put("properties", JSONObject().apply {
                    put("action", JSONObject().apply {
                        put("type", "string")
                        put("enum", JSONArray().put("delegate").put("render"))
                        put(
                            "description",
                            "delegate = perform a real-world action via the assistant (OpenClaw); render = show widget cards in the field of view.",
                        )
                    })
                    put("task", JSONObject().apply {
                        put("type", "string")
                        put(
                            "description",
                            "For action=delegate: a clear, detailed description of what to do, with all relevant context (names, content, platforms, quantities, etc.).",
                        )
                    })
                    put("widgets", JSONObject().apply {
                        put("type", "array")
                        put(
                            "description",
                            "For action=render: the widget cards to show, top to bottom. Each render replaces whatever is currently shown.",
                        )
                        put("items", widgetItemSchema())
                    })
                })
                put("required", JSONArray().put("action"))
            })
            put("behavior", "BLOCKING")
        }
    }

    private fun widgetItemSchema(): JSONObject {
        return JSONObject().apply {
            put("type", "object")
            put("properties", JSONObject().apply {
                put("type", JSONObject().apply {
                    put("type", "string")
                    put("enum", JSONArray().put("text").put("image").put("table"))
                    put("description", "The widget type.")
                })
                put("title", JSONObject().apply {
                    put("type", "string")
                    put("description", "Optional heading shown at the top of the widget.")
                })
                put("body", JSONObject().apply {
                    put("type", "string")
                    put("description", "For type=text: the paragraph of text to show.")
                })
                put("imageKind", JSONObject().apply {
                    put("type", "string")
                    put("enum", JSONArray().put("map").put("gallery").put("calendar"))
                    put(
                        "description",
                        "For type=image: which fixed image to show — map, gallery, or calendar.",
                    )
                })
                put("caption", JSONObject().apply {
                    put("type", "string")
                    put("description", "For type=image: a short caption under the image.")
                })
                put("columns", JSONObject().apply {
                    put("type", "array")
                    put("items", JSONObject().put("type", "string"))
                    put("description", "For type=table: optional column header labels.")
                })
                put("rows", JSONObject().apply {
                    put("type", "array")
                    put(
                        "description",
                        "For type=table: the rows; each row is an array of cell strings.",
                    )
                    put(
                        "items",
                        JSONObject().apply {
                            put("type", "array")
                            put("items", JSONObject().put("type", "string"))
                        },
                    )
                })
            })
            put("required", JSONArray().put("type"))
        }
    }
}
